File: algo/problems/0001-two-sum/js/twoSum.js
/**
 * 一次遍历 + Map：seen.get(值)=下标，命中 target-x 即返回。O(n)/O(n)。
 * @param {number[]} nums
 * @param {number} target
 * @returns {number[]}
 */
function twoSum(nums, target) {
  const seen = new Map();
  for (let i = 0; i < nums.length; i++) {
    const need = target - nums[i];
    if (seen.has(need)) return [seen.get(need), i];
    seen.set(nums[i], i);
  }
  return [];
}

module.exports = { twoSum };

File: algo/problems/0001-two-sum/go/twosum.go
package twosum

// TwoSum 一次遍历 + map：seen[值]=下标，命中 target-x 即返回。O(n)/O(n)。
func TwoSum(nums []int, target int) []int {
	seen := make(map[int]int, len(nums))
	for i, x := range nums {
		if j, ok := seen[target-x]; ok {
			return []int{j, i}
		}
		seen[x] = i
	}
	return nil
}
